Doc/library/asyncio-future.rst

```diff
@@ -49,6 +49,9 @@ Future Functions
 
    If *obj* is neither of the above a :exc:`TypeError` is raised.
 
+   An optional keyword-only *context* argument allows specifying a
+   custom :class:`contextvars.Context` to use when creating a new task.
+
```

Comment:
Mention that the current context is used if none is specified.

Reply:
From what I can tell, `ensure_future()` sets no context at all if you pass in `None` to it. It calls `create_task()` on the current event loop with no context argument in that case. What happens from there would depend on the event loop being used.

I do see code in `Task()` constructor which calls `contextvars.copy_context()` if you pass `None` in as a context and that's what event loops usually call.  I'm hesitant to claim that behavior at the `ensure_future()` level, though, given the multiple layers in between and the possibility for the event loops to do something different.

Looking at `Runner.run()`, it passes in the runner's default context if no context is specified when `Runner.run()` is called, and it is documented as such. That default context seems to be set to a copy of the context active at the time the Runner object is first used (during  its lazy init). That then gets passed explicitly to `ensure_future()` if no context is passed into `Runner.run()`.

Note that the runner's default context may not always be the same as the context which is active at the time `Runner.run()` is called.

Reply:
You're entirely right. I incorrectly assumed that it would be like `Runner.run()` but that's not what the code actually does. Maybe we could expand the docs by saying:

```diff
-custom :class:`contextvars.Context` to use when creating a new task.
+custom :class:`contextvars.Context` to use when creating a new task
+via :meth:`loop.create_task <asyncio.loop.create_task>`.
```

Reply:
I don't think the loop version of the create_task() method is a public function. There's only asyncio.create_task() (and more recently TaskGroup.create_task), but those are not used by Runner.

Reply:
Then I really wonder how we can teach the None argument in that case. I don't have any good suggestion but we can also decide to be deliberately imprecise... I'll leave it to another reviewer :/

Reply:
I looked this over again, and it turns out that the same indirection happening with `ensure_future()` calling into the internal `create_task()` on the event loop also happens in `asyncio.create_task()`, and yet it documents context as:

```
   An optional keyword-only *context* argument allows specifying a
   custom :class:`contextvars.Context` for the *coro* to run in.
   The current context copy is created when no *context* is provided.
```

Given that both of these are likely to call the constructor for `Task` which does the actual copy and documents that:

```
   An optional keyword-only *context* argument allows specifying a
   custom :class:`contextvars.Context` for the *coro* to run in.
   If no *context* is provided, the Task copies the current context
   and later runs its coroutine in the copied context.
```

So, perhaps it's ok after all to document context in `ensure_future()` as something like:

```
   An optional keyword-only *context* argument allows specifying a
   custom :class:`contextvars.Context` to use when creating a new task.
   A copy of the current context is used when no *context* is provided.
```

There's a chance that a custom event loop could fail to pass the context through properly and change this behavior but that's fairly unlikely (and might in fact be considered a bug).